entity-framework/core/modeling/custom-function-mapping.md

```diff
@@ -0,0 +1,201 @@
+
+[SqlExpressionFactory](dotnet/api/microsoft.entityframeworkcore.query.sqlexpressionfactory) can be used to construct `SqlExpression` tree.
+
+Once we define the function, it can be used in the query. Instead of calling database function, EF Core will translate the method body directly into SQL.
+
+[!code-csharp[Main](../../../samples/core/Modeling/CustomFunctionMapping/Program.cs#HasTranslationQuery)]
+
+Produces the following SQL:
+
+```sql
+SELECT [p].[PostId], [p].[BlogId], [p].[Content], [p].[Rating], [p].[Title]
+FROM [Posts] AS [p]
+WHERE [p].[PostId] < ABS([p].[BlogId] - 3)
+```
+
+## Mapping DbSet to a Table-Valued Function
+
+It's also possible to map a `DbSet` of entities to a Table-Valued function instead of a table in the database. To illustrate this lets define another entity that represents blog with multiple posts. In the example, the entity is [keyless](core/modeling/keyless-entity-types), but it doesn't have to be.
+
+[!code-csharp[Main](../../../samples/core/Modeling/CustomFunctionMapping/BlogWithMultiplePosts.cs#Entity)]
+
+Next, create the following Table-Valued Function on the database, which returns only blogs with multiple posts as well as the number of posts associated with each of these blogs:
+
+```sql
+CREATE FUNCTION dbo.BlogsWithMultiplePosts()
+RETURNS @blogs TABLE
+(
+    Rating int,
+    Url nvarchar(max),
+    PostCount int not null
+)
+AS
+BEGIN
+    INSERT INTO @blogs
+    SELECT b.Rating, b.Url, COUNT(p.BlogId)
+    FROM Blogs AS b
+    JOIN Posts AS p ON b.BlogId = p.BlogId
+    GROUP BY b.Rating, b.Url
+    HAVING COUNT(p.BlogId) > 1
+
+    RETURN
+END
+```
+
+Now, the `DbSet<BlogWithMultiplePost>` can be mapped to this function in a following way:
+
+[!code-csharp[Main](../../../samples/core/Modeling/CustomFunctionMapping/BloggingContext.cs#QueryableFunctionConfigurationToFunction)]
+
+> [!NOTE]
+> In order to map DbSet to a Table-Valued Function the function must be parameterless. Also, names of the entity properties should match the names of the columns returned by the TVF. Any discrepancies can be configured using `HasColumnName` method, just like mapping to a table.
+
+When the set is mapped to a Table-Valued function, the query:
+
+[!code-csharp[Main](../../../samples/core/Modeling/CustomFunctionMapping/Program.cs#ToFunctionQuery)]
+
+Produces the following SQL:
+
+```sql
+SELECT [b].[Url], [b].[PostCount]
+FROM [dbo].[BlogsWithMultiplePosts]() AS [b]
+WHERE [b].[Rating] > 3
+```
+
+## Mapping Queryable function to a Table-Valued Function
+
+EF Core also supports mapping to Table-Valued Function using a user-defined CLR function returning `IQueryable` of entity types. This allows EF Core to use Table-Valued Function with parameters. Process is similar to mapping a scalar user-defined function to a SQL function. We need a Table-Valued function on the database, CLR function that will be used in the LINQ queries and mapping between the two.
```

Comment:
"a SQL" or "an SQL"?

Reply:
Not this discussion again :scream: :scream: :scream: :rofl: